Algorithms/Strings/ZblockSubstringSearch.cs

```diff
@@ -0,0 +1,70 @@
+namespace Algorithms.Strings
+{
+    /// <summary>Implementation Z-block substring search.
+    /// </summary>
+    public static class ZblockSubstringSearch
+    {
+        /// <summary>
+        ///     This algorithm finds all occurrences of a pattern in a text in linear time - O(m+n).
+        /// </summary>
+        public static int FindSubstring(string pattern, string text)
+        {
+            var concatStr = $"{pattern}${text}";
+            var patternLength = pattern.Length;
+            var n = concatStr.Length;
+            var zArray = new int[n];
+
+            var left = 0;
+            var right = 0;
+
+            for(var i = 1; i < n; i++)
+            {
+                if(i > right)
+                {
+                    left = i;
+                    right = i;
+                    right = ComputeNewRightValue(concatStr, n, left, right);
```

Comment:
The first assignment is redundant
```suggestion
                    right = ComputeNewRightValue(concatStr, n, left, i);
```